kernel/include/mmu.h

```diff
@@ -98,68 +105,89 @@
  */
 struct z_page_frame {
 	union {
-		/* If mapped, virtual address this page is mapped to */
-		void *addr;
-
-		/* If unmapped and available, free pages list membership. */
-		sys_snode_t node;
+		/*
+		 * If mapped, Z_PAGE_FRAME_* flags and virtual address
+		 * this page is mapped to.
+		 */
+		uintptr_t va_and_flags;
+
+		/*
+		 * If unmapped and available, free pages list membership
+		 * with the Z_PAGE_FRAME_FREE flag.
+		 */
+		sys_sfnode_t node;
 	};
 
-	/* Z_PAGE_FRAME_* flags */
-	uint8_t flags;
-
-	/* TODO: Backing store and eviction algorithms may both need to
-	 * introduce custom members for accounting purposes. Come up with
-	 * a layer of abstraction for this. They may also want additional
-	 * flags bits which shouldn't clobber each other. At all costs
-	 * the total size of struct z_page_frame must be minimized.
+	/* Backing store and eviction algorithms may both need to
+	 * require additional per-frame custom data for accounting purposes.
+	 * They should declare their own array with indices matching
+	 * z_page_frames[] ones whenever possible.
+	 * They may also want additional flags bits that could be stored here
+	 * and they shouldn't clobber each other. At all costs the total
+	 * size of struct z_page_frame must be minimized.
 	 */
+};
 
-	/* On Xtensa we can't pack this struct because of the memory alignment.
-	 */
-#ifdef CONFIG_XTENSA
-} __aligned(4);
-#else
-} __packed;
-#endif /* CONFIG_XTENSA */
+/* Note: this must be false for the other flag bits to be valid */
+static inline bool z_page_frame_is_free(struct z_page_frame *pf)
+{
+	return (pf->va_and_flags & Z_PAGE_FRAME_FREE) != 0U;
+}
 
 static inline bool z_page_frame_is_pinned(struct z_page_frame *pf)
 {
-	return (pf->flags & Z_PAGE_FRAME_PINNED) != 0U;
+	return (pf->va_and_flags & Z_PAGE_FRAME_PINNED) != 0U;
 }
 
 static inline bool z_page_frame_is_reserved(struct z_page_frame *pf)
 {
-	return (pf->flags & Z_PAGE_FRAME_RESERVED) != 0U;
+	return (pf->va_and_flags & Z_PAGE_FRAME_RESERVED) != 0U;
 }
 
 static inline bool z_page_frame_is_mapped(struct z_page_frame *pf)
 {
-	return (pf->flags & Z_PAGE_FRAME_MAPPED) != 0U;
+	return (pf->va_and_flags & Z_PAGE_FRAME_MAPPED) != 0U;
 }
 
 static inline bool z_page_frame_is_busy(struct z_page_frame *pf)
 {
-	return (pf->flags & Z_PAGE_FRAME_BUSY) != 0U;
+	return (pf->va_and_flags & Z_PAGE_FRAME_BUSY) != 0U;
 }
 
 static inline bool z_page_frame_is_backed(struct z_page_frame *pf)
 {
-	return (pf->flags & Z_PAGE_FRAME_BACKED) != 0U;
+	return (pf->va_and_flags & Z_PAGE_FRAME_BACKED) != 0U;
 }
 
 static inline bool z_page_frame_is_evictable(struct z_page_frame *pf)
 {
-	return (!z_page_frame_is_reserved(pf) && z_page_frame_is_mapped(pf) &&
-		!z_page_frame_is_pinned(pf) && !z_page_frame_is_busy(pf));
+	return (!z_page_frame_is_free(pf) &&
+		!z_page_frame_is_reserved(pf) &&
+		z_page_frame_is_mapped(pf) &&
+		!z_page_frame_is_pinned(pf) &&
+		!z_page_frame_is_busy(pf));
```

Comment:
It might be easier for the compiler to omit better code by grouping the negated ANDed terms into a single negation of the ORed terms.

The ORs might be able to short circuit as well.

If there was a likelihood ordering, that might be faster on average as well.

This is a very weak suggestion based on my preconceived notion of what instructions a compiler might omit rather than actual disassembly though.

Godbolt is clicks away, but I am on mobile and lazy atm 🤷🏼‍♂️

Feel free to ignore.

```suggestion
	return (z_page_frame_is_mapped(pf) &&
		!(z_page_frame_is_free(pf) ||
		z_page_frame_is_reserved(pf) ||
		z_page_frame_is_pinned(pf) ||
		z_page_frame_is_busy(pf)));
```

Reply:
Or go further and change the sense of "pinned" to "unpinned", then it becomes equivalent to a `flags & mask == 0` computation.